Adding Selection Support

haven1433 opened the pull request:
* added properties for selection start / selection end
* added commands for moving the selection / scrolling
* added key bindings for those commands
* added logic for drawing the selection / mouse control of selection
* added unit tests selection / the interaction between selection/scrolling
* visual style for main view via Solarized
* ReadMe document explaining the vision of the product.

Note to reviewers: Scrolling is still line-based: the ScrollValue is an integer that represents what line of the document you are on. The selection is based on what is visible in the ViewPort: a selection of (0,0) is the upper left cell. This scheme was chosen to minimize the view logic: the view just asks if a given visible cell is selected and can draw accordingly, instead of needing to do any math to map between selected data and selected cells.

rtzoeller commented:
What should the behavior be if I ctrl-left arrow to the point where only one element is in the top row, and then press it again? Currently my selection moves to the right, while the data doesn't shift. While this behavior isn't wrong, it isn't what I expected (the data to shift over by one, causing a new line to appear and the values to move up to fill the newly created blank line).

haven1433 commented:
Good catch! I added a new test (ScrollingBeforeStartOfDataMovesSelectionOnlyWhenDataMoves) that captures this edge case.

rtzoeller commented:
This is a minor thing that probably isn't worth addressing, but if I load an entirely empty file moving around with ctrl-arrow keys behaves in a really weird way. Steps to reproduce below.

 - Open an empty file
 - Click somewhere in the middle of the window, causing another empty element to be selected
 - At this point, both ctrl-up or ctrl-down will result in the same movement (need to restart to try the other one)
 - Holding ctrl-left or ctrl-right will result in the selection moving, but stopping moving at the originally selected point (after wrapping once)

haven1433 commented:
Little issues are the ones that sneak in and make the program feel "quirky" to the user. I'd rather just disallow bad behavior.

Besides, you finding this issue showed me another issue: it was possible for the scroll maximum to be smaller than the scroll minimum. I added two tests and adjusted the code accordingly: empty files should behave a bit better now.

I still haven't decided whether I should cap the selection to the start/end of the file instead of allowing the user to select cells that are out of range of the data. I'll think a bit more on that and consider adding a test. But since I never want to deal with the selection being null, at least the first cell must always be selectable (in the case of empty data).
